lib/orbit/transformable.js

```diff
@@ -1,121 +1,118 @@
+    data: transform,
+    process: function() {
+      return processTransform.call(_this, transform);
+    }
+  });
+
+  return action.complete;
 }
 
-function queueTransformation(transformation) {
+function queueTransformResult(transform, result) {
   var _this = this;
 
-  var processor = this._transformationQueue.push({
-    data: transformation,
+  var action = this._transformQueue.push({
+    data: transform,
     process: function() {
-      return transformation.process();
+      return settleTransform.call(_this, transform, result);
     }
   });
 
-  return processor;
+  return action.complete;
 }
 
 var Transformable = {
   extend: function(object, actions) {
     if (object._transformable === undefined) {
       object._transformable = true;
-      object._transformationQueue = new ActionQueue();
+      object._transformQueue = new ActionQueue();
+      object._transformLog = {};
 
       Evented.extend(object);
 
-      object.didTransform = function(operation, inverse) {
-        var normalized = normalize(operation);
-        var transformation = transformationFor.call(this, normalized);
-        if (transformation) {
-          // console.log('Transformable#didTransform - matching transformation found', this.id, normalized, inverse);
-          transformation.pushCompletedOperation(normalized, inverse);
-
-        } else {
-          // console.log('Transformable#didTransform - createTransformation', this.id, normalized, inverse);
-          transformation = new Transformation(this);
-          transformation.pushCompletedOperation(normalized, inverse);
-          queueTransformation.call(this, transformation);
-        }
+      object.didTransform = function() {
+        Orbit.deprecate('`didTransform` has been deprecated. Please call `transformed` instead and supply a TransformResult as an argument.', true);
       };
 
-      object.currentTransformation = function() {
-        if (this._transformationQueue.current) return this._transformationQueue.current.data;
-      };
+      object.transformed = function(result) {
+        assert('Call `transformed` with an instanceof a TransformResult.', result instanceof TransformResult);
 
-      object.transform = function(operation) {
-        var normalized = normalize(operation);
-        var transformation = transformationFor.call(this, normalized);
-        var action;
+        var transform = new Transform(result.operations);
 
-        if (transformation) {
-          // console.log('transform - matching transformation found', this.id, normalized);
-          action = transformation.pushOperation(normalized);
+        // console.log('transformed - queued', this.id, transform, result);
+        return queueTransformResult.call(this, transform, result);
+      };
 
-          if (isArray(action)) {
-            return action[action.length - 1].complete;
-          } else {
-            return action.complete;
-          }
+      object.transform = function(transform) {
```

Comment:
I currently use an optimistic configuration, so orbit-firebase is hooked up with a non-blocking connector and in addition transforms respond immediately. This allows updates to firebase to occur faster because they can be queued up behind each other on the websocket rather than having to wait for each one to complete. Firebase guarantees that transformations will be applied in the requested sequence so they're applied (and received) in the same sequence as the calls to source.transform(). 

Regarding the pessimistic scenario - I do actually capture operations that the client has transmitted and received back again. I filter those operations out so that source only emits operations that originated in other clients (this prevents churning in certain situtations). I could hook into this to implement the pessimistic scenario, but having an optimistic UI is a much snappier experience and in practice there aren't any errors unless the network connection drops out.

The key difference with ember-data is that there's no sequencing of operations, they all happen in parallel. In that sense the records are more directly connected to firebase and because there're no other sources there's no need to worry about sequencing.